File: src/topic_demo/topic_demo/client_future.py
# =============================================================================
# 【学习 · Future 等待客户端】client_future.py
#
# 机制：
#   无长期 spin。wait_for_service → call_async → spin_until_future_complete
#   → result() → 退出。
#
# 对照：my_topic.py = 常驻 spin + service_is_ready + add_done_callback
#
# 运行：先 ros2 run topic_demo py_topic，再本程序。
# =============================================================================

import sys

import rclpy
from example_interfaces.srv import AddTwoInts


def main(args=None):
    rclpy.init(args=args)
    node = rclpy.create_node('client_future')
    client = node.create_client(AddTwoInts, 'add_two_ints')

    # =========================================================================
    # wait_for_service(timeout_sec)
    #   阻塞最多 timeout_sec；服务出现 True，超时 False
    #   适合启动阶段 while 轮询
    # service_is_ready()
    #   立刻返回当前是否可用，不阻塞
    #   适合已在 spin 的 timer/回调里（见 my_topic.py）
    # =========================================================================
    while not client.wait_for_service(timeout_sec=1.0):
        if not rclpy.ok():
            node.get_logger().error('Interrupted while waiting for service.')
            node.destroy_node()
            rclpy.shutdown()
            return 1
        node.get_logger().info('wait_for_service(1.0) timed out, retry...')
    node.get_logger().info('service ready (wait_for_service returned True)')

    request = AddTwoInts.Request()
    request.a = 41
    request.b = 1

    # =========================================================================
    # 【写法 A · Future】
    # call_async(request)  —— 非阻塞发送
    # spin_until_future_complete(node, future, timeout_sec)
    #   —— 本程序唯一的 spin：边处理回调边等 Future
    # future.done() / future.result()
    #   —— 判断完成；取结果（未完成时 result 可能抛异常，故先 done）
    #
    # 注意：外面若已有 rclpy.spin(node)，不要再用本写法
    # =========================================================================
    future = client.call_async(request)
    rclpy.spin_until_future_complete(node, future, timeout_sec=5.0)

    if future.done():
        try:
            result = future.result()
        except Exception as e:
            node.get_logger().error('Service call failed: %r' % (e,))
            node.destroy_node()
            rclpy.shutdown()
            return 1
        node.get_logger().info(
            '[future] %d + %d = %d' % (request.a, request.b, result.sum))
    else:
        node.get_logger().error('spin_until_future_complete timeout')
        node.destroy_node()
        rclpy.shutdown()
        return 1

    node.destroy_node()
    rclpy.shutdown()
    return 0


if __name__ == '__main__':
    sys.exit(main())

File: src/my_package/src/client_future.cpp
// =============================================================================
// 【学习 · Future 等待客户端】client_future.cpp
//
// 机制：
//   本程序没有长期 rclcpp::spin。
//   wait_for_service 等服务上线 → async_send_request → spin_until_future_complete
//   等结果 → get → 退出。
//
// 对照：
//   my_node.cpp = 常驻 spin + service_is_ready + callback
//
// 运行：先 ros2 run my_package my_node，再本程序。
// =============================================================================

#include <chrono>
#include <cinttypes>
#include <memory>

#include "example_interfaces/srv/add_two_ints.hpp"
#include "rclcpp/rclcpp.hpp"

using namespace std::chrono_literals;
using AddTwoInts = example_interfaces::srv::AddTwoInts;

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  // 脚本式节点：不继承类也可以 Node::make_shared
  auto node = rclcpp::Node::make_shared("client_future");
  auto client = node->create_client<AddTwoInts>("add_two_ints");

  // =========================================================================
  // wait_for_service(timeout)
  //   - 阻塞最多 timeout，服务出现返回 true，超时返回 false
  //   - 适合：启动阶段「一直等到能调」
  // service_is_ready()
  //   - 立刻返回当前是否可用，不阻塞
  //   - 适合：已在 spin 的 timer/回调里「查一下再决定发不发」
  // 本文件是脚本：用 while + wait_for_service 轮询
  // =========================================================================
  while (!client->wait_for_service(1s)) {
    if (!rclcpp::ok()) {
      RCLCPP_ERROR(node->get_logger(), "Interrupted while waiting for service.");
      return 1;
    }
    RCLCPP_INFO(node->get_logger(), "wait_for_service(1s) timed out, retry...");
  }
  RCLCPP_INFO(node->get_logger(), "service ready (wait_for_service returned true)");

  auto request = std::make_shared<AddTwoInts::Request>();
  request->a = 41;
  request->b = 1;

  // =========================================================================
  // 【写法 A · Future】
  // async_send_request(req)  —— 非阻塞发送，返回 Future
  // spin_until_future_complete(node, future, timeout)
  //   —— 本程序唯一的 spin：边处理 ROS 回调，边等 Future
  //   —— 返回 SUCCESS / TIMEOUT / INTERRUPTED
  // future.get()             —— SUCCESS 后再取结果；未完成时 get 会阻塞
  // remove_pending_request   —— 失败时清掉未完成请求，避免泄漏
  //
  // 注意：外面若已有 spin(node)，不要再用本写法（见 my_node.cpp）
  // =========================================================================
  auto result_future = client->async_send_request(request);

  auto ret = rclcpp::spin_until_future_complete(node, result_future, 5s);
  if (ret != rclcpp::FutureReturnCode::SUCCESS) {
    RCLCPP_ERROR(node->get_logger(), "spin_until_future_complete failed");
    client->remove_pending_request(result_future);
    rclcpp::shutdown();
    return 1;
  }

  auto result = result_future.get();
  RCLCPP_INFO(
    node->get_logger(),
    "[future] %" PRId64 " + %" PRId64 " = %" PRId64,
    request->a, request->b, result->sum);

  rclcpp::shutdown();
  return 0;
}
